File: qrs-mobile-app/lib/stores.ts
/**
 * Persistent stores for the qrs-core runtime, backed by AsyncStorage.
 *
 * Only the stores a *verification* app needs are persisted: certificates (TCerts),
 * trust, revocation and documents. Private/public keys keep the in-memory defaults
 * (verification does not need to sign).
 */
import AsyncStorage from '@react-native-async-storage/async-storage';
import { parseSignedObject, toBase64Url, fromBase64Url } from 'qrs-core';
import type {
  AttestationRecord,
  BlockEntry,
  ICertificateStore,
  IDocumentStore,
  IEndpointConfigStore,
  IRevocationStore,
  ITrustStore,
  RevocationEntry,
  SdocStatementEntry,
} from 'qrs-core';

const PREFIX = 'qrs.';

function key(name: string): string {
  return PREFIX + name;
}

async function readJson<T>(name: string, fallback: T): Promise<T> {
  try {
    const raw = await AsyncStorage.getItem(key(name));
    return raw ? (JSON.parse(raw) as T) : fallback;
  } catch {
    return fallback;
  }
}

async function writeJson(name: string, value: unknown): Promise<void> {
  await AsyncStorage.setItem(key(name), JSON.stringify(value));
}

const b64 = (bytes: Uint8Array): string => toBase64Url(bytes);
const unb64 = (s: string): Uint8Array => fromBase64Url(s);

export const certificateStore: ICertificateStore = {
  async save(tcertId, bytes) {
    const all = await readJson<Array<{ id: string; bytes: string }>>('certificates', []);
    const idx = all.findIndex((c) => c.id === tcertId);
    const entry = { id: tcertId, bytes: b64(bytes) };
    if (idx >= 0) all[idx] = entry;
    else all.push(entry);
    await writeJson('certificates', all);
  },
  async get(tcertId) {
    const all = await readJson<Array<{ id: string; bytes: string }>>('certificates', []);
    const hit = all.find((c) => c.id === tcertId);
    return hit ? unb64(hit.bytes) : null;
  },
  async findByKeyId(keyId) {
    const all = await readJson<Array<{ id: string; bytes: string }>>('certificates', []);
    const out: Array<{ tcertId: string; bytes: Uint8Array }> = [];
    for (const c of all) {
      const bytes = unb64(c.bytes);
      try {
        const parsed = parseSignedObject(bytes);
        if (parsed.type === 'tcert' && parsed.signerKeyId === keyId) out.push({ tcertId: c.id, bytes });
      } catch {
        /* skip malformed */
      }
    }
    return out;
  },
  async all() {
    const all = await readJson<Array<{ id: string; bytes: string }>>('certificates', []);
    return all.map((c) => ({ tcertId: c.id, bytes: unb64(c.bytes) }));
  },
  async remove(tcertId) {
    const all = await readJson<Array<{ id: string; bytes: string }>>('certificates', []);
    await writeJson(
      'certificates',
      all.filter((c) => c.id !== tcertId),
    );
  },
};

export const documentStore: IDocumentStore = {
  async save(sdocId, bytes) {
    const all = await readJson<Array<{ id: string; bytes: string }>>('documents', []);
    const idx = all.findIndex((d) => d.id === sdocId);
    const entry = { id: sdocId, bytes: b64(bytes) };
    if (idx >= 0) all[idx] = entry;
    else all.push(entry);
    await writeJson('documents', all);
  },
  async get(sdocId) {
    const all = await readJson<Array<{ id: string; bytes: string }>>('documents', []);
    const hit = all.find((d) => d.id === sdocId);
    return hit ? unb64(hit.bytes) : null;
  },
  async all() {
    const all = await readJson<Array<{ id: string; bytes: string }>>('documents', []);
    return all.map((d) => ({ sdocId: d.id, bytes: unb64(d.bytes) }));
  },
  async remove(sdocId) {
    const all = await readJson<Array<{ id: string; bytes: string }>>('documents', []);
    await writeJson(
      'documents',
      all.filter((d) => d.id !== sdocId),
    );
  },
};

export const revocationStore: IRevocationStore = {
  async addRevokedTcert(tcertId, entry) {
    const map = await readJson<Record<string, RevocationEntry[]>>('revoked-tcerts', {});
    const list = map[tcertId] ?? [];
    if (entry.byKeyId) {
      const idx = list.findIndex((e) => e.byKeyId === entry.byKeyId);
      if (idx >= 0) list[idx] = entry;
      else list.push(entry);
    } else {
      list.push(entry);
    }
    map[tcertId] = list;
    await writeJson('revoked-tcerts', map);
  },
  async getRevokedTcert(tcertId) {
    const map = await readJson<Record<string, RevocationEntry[]>>('revoked-tcerts', {});
    const list = map[tcertId] ?? [];
    if (list.length === 0) return null;
    const sorted = [...list].sort((a, b) => {
      const sev = (t: RevocationEntry['type']): number => (t === 'retrospective' ? 1 : 0);
      return sev(b.type) - sev(a.type) || b.issuedAt - a.issuedAt;
    });
    return sorted[0] ?? null;
  },
  async getRevokedTcertEntries(tcertId) {
    const map = await readJson<Record<string, RevocationEntry[]>>('revoked-tcerts', {});
    return [...(map[tcertId] ?? [])];
  },
  async listRevokedTcert() {
    const map = await readJson<Record<string, RevocationEntry[]>>('revoked-tcerts', {});
    return Object.entries(map).flatMap(([tcertId, list]) =>
      list.map((entry) => ({ tcertId, entry }))
    );
  },
  async addRevokedKey(keyId, entry) {
    const map = await readJson<Record<string, RevocationEntry>>('revoked-keys', {});
    map[keyId] = entry;
    await writeJson('revoked-keys', map);
  },
  async getRevokedKey(keyId) {
    const map = await readJson<Record<string, RevocationEntry>>('revoked-keys', {});
    return map[keyId] ?? null;
  },
  async listRevokedKey() {
    const map = await readJson<Record<string, RevocationEntry>>('revoked-keys', {});
    return Object.entries(map).map(([keyId, entry]) => ({ keyId, entry }));
  },
  async addRevokedAttestation(targetTcertId, caTcertId, entry) {
    const map = await readJson<Record<string, RevocationEntry>>('revoked-attestations', {});
    map[`${targetTcertId}|${caTcertId}`] = entry;
    await writeJson('revoked-attestations', map);
  },
  async getRevokedAttestation(targetTcertId, caTcertId) {
    const map = await readJson<Record<string, RevocationEntry>>('revoked-attestations', {});
    return map[`${targetTcertId}|${caTcertId}`] ?? null;
  },
  async listRevokedAttestation() {
    const map = await readJson<Record<string, RevocationEntry>>('revoked-attestations', {});
    return Object.entries(map).map(([key, entry]) => {
      const [targetTcertId, caTcertId] = key.split('|');
      return { targetTcertId: targetTcertId!, caTcertId: caTcertId!, entry };
    });
  },
  async addBlockedSdoc(sdocId, entry) {
    const map = await readJson<Record<string, BlockEntry>>('blocked-sdocs', {});
    map[sdocId] = entry;
    await writeJson('blocked-sdocs', map);
    const history = await readJson<Array<{ sdocId: string; entry: Omit<SdocStatementEntry, 'statementBytes'> & { statementBytes?: string } }>>('sdoc-statements', []);
    history.push({ sdocId, entry: { ...entry, action: 'blockSdoc', statementBytes: entry.statementBytes ? b64(entry.statementBytes) : undefined } });
    await writeJson('sdoc-statements', history);
  },
  async getBlockedSdoc(sdocId) {
    const map = await readJson<Record<string, BlockEntry>>('blocked-sdocs', {});
    return map[sdocId] ?? null;
  },
  async listBlockedSdoc() {
    const map = await readJson<Record<string, BlockEntry>>('blocked-sdocs', {});
    return Object.entries(map).map(([sdocId, entry]) => ({ sdocId, entry }));
  },
  async removeBlockedSdoc(sdocId, entry) {
    const map = await readJson<Record<string, BlockEntry>>('blocked-sdocs', {});
    delete map[sdocId];
    await writeJson('blocked-sdocs', map);
    if (entry) {
      const history = await readJson<Array<{ sdocId: string; entry: Omit<SdocStatementEntry, 'statementBytes'> & { statementBytes?: string } }>>('sdoc-statements', []);
      history.push({ sdocId, entry: { ...entry, action: 'unblockSdoc', statementBytes: entry.statementBytes ? b64(entry.statementBytes) : undefined } });
      await writeJson('sdoc-statements', history);
    }
  },
  async listSdocStatements() {
    const history = await readJson<Array<{ sdocId: string; entry: Omit<SdocStatementEntry, 'statementBytes'> & { statementBytes?: string } }>>('sdoc-statements', []);
    return history.map(({ sdocId, entry }) => ({
      sdocId,
      entry: { ...entry, statementBytes: entry.statementBytes ? unb64(entry.statementBytes) : undefined } as SdocStatementEntry,
    }));
  },
};

type StoredAttestation = Omit<AttestationRecord, 'statementBytes'> & { statementBytes: string };

export const trustStore: ITrustStore = {
  async addPinned(tcertId) {
    const list = await readJson<string[]>('pinned', []);
    if (!list.includes(tcertId)) {
      list.push(tcertId);
      await writeJson('pinned', list);
    }
  },
  async removePinned(tcertId) {
    const list = await readJson<string[]>('pinned', []);
    await writeJson(
      'pinned',
      list.filter((id) => id !== tcertId),
    );
  },
  async isPinned(tcertId) {
    const list = await readJson<string[]>('pinned', []);
    return list.includes(tcertId);
  },
  async listPinned() {
    return readJson<string[]>('pinned', []);
  },
  async addCa(tcertId) {
    const list = await readJson<string[]>('cas', []);
    if (!list.includes(tcertId)) {
      list.push(tcertId);
      await writeJson('cas', list);
    }
  },
  async removeCa(tcertId) {
    const list = await readJson<string[]>('cas', []);
    await writeJson(
      'cas',
      list.filter((id) => id !== tcertId),
    );
  },
  async isCa(tcertId) {
    const list = await readJson<string[]>('cas', []);
    return list.includes(tcertId);
  },
  async listCa() {
    return readJson<string[]>('cas', []);
  },
  async addAttestation(record) {
    const list = await readJson<StoredAttestation[]>('attestations', []);
    const entry: StoredAttestation = {
      targetTcertId: record.targetTcertId,
      caKeyId: record.caKeyId,
      caTcertId: record.caTcertId,
      tcertHash: record.tcertHash,
      claims: record.claims,
      issuedAt: record.issuedAt,
      statementBytes: b64(record.statementBytes),
    };
    // One attestation per (target, CA): re-importing the same CA's attestation
    // must replace, not duplicate. Mirrors qrs-core's FileTrustStore behavior
    // and keeps the per-CA result pages unique (no duplicate React keys).
    const idx = list.findIndex(
      (a) => a.targetTcertId === record.targetTcertId && a.caTcertId === record.caTcertId
    );
    if (idx >= 0) list[idx] = entry;
    else list.push(entry);
    await writeJson('attestations', list);
  },
  async getAttestations(targetTcertId) {
    const list = await readJson<StoredAttestation[]>('attestations', []);
    // Defensive dedup by (target, CA) so stale duplicates never surface.
    const seen = new Set<string>();
    const deduped = list.filter((a) => {
      if (a.targetTcertId !== targetTcertId) return false;
      const k = a.caTcertId;
      if (seen.has(k)) return false;
      seen.add(k);
      return true;
    });
    return deduped.map((a) => ({ ...a, statementBytes: unb64(a.statementBytes) }));
  },
  async listAttestations() {
    const list = await readJson<StoredAttestation[]>('attestations', []);
    return list.map((record) => ({ ...record, statementBytes: fromBase64Url(record.statementBytes) }));
  },
  async addDistrust(tcertId) {
    const list = await readJson<string[]>('distrusted', []);
    if (!list.includes(tcertId)) {
      list.push(tcertId);
      await writeJson('distrusted', list);
    }
  },
  async removeDistrust(tcertId) {
    const list = await readJson<string[]>('distrusted', []);
    await writeJson(
      'distrusted',
      list.filter((id) => id !== tcertId),
    );
  },
  async isDistrusted(tcertId) {
    const list = await readJson<string[]>('distrusted', []);
    return list.includes(tcertId);
  },
};

export async function listArchivedTcertIds(): Promise<string[]> {
  return readJson<string[]>('archived-tcerts', []);
}

export async function setTcertArchived(tcertId: string, archived: boolean): Promise<void> {
  const ids = await listArchivedTcertIds();
  await writeJson('archived-tcerts', archived ? [...new Set([...ids, tcertId])] : ids.filter((id) => id !== tcertId));
}

/** Delete a certificate and the trust graph that depends exclusively on it. */
export async function removeTcertCascade(tcertId: string): Promise<{ removedTcertIds: string[] }> {
  const keyId = tcertId.slice(0, tcertId.lastIndexOf(':'));
  const attestations = await readJson<StoredAttestation[]>('attestations', []);
  const issuedByRemoved = attestations.filter((a) => a.caTcertId === tcertId);
  const remainingAttestations = attestations.filter((a) => a.caTcertId !== tcertId);
  const pinned = await readJson<string[]>('pinned', []);
  const orphanedTargets = issuedByRemoved
    .map((a) => a.targetTcertId)
    .filter((targetId, index, all) => all.indexOf(targetId) === index)
    .filter((targetId) => !pinned.includes(targetId) && !remainingAttestations.some((a) => a.targetTcertId === targetId));
  const removedTcertIds = [tcertId, ...orphanedTargets];
  const removed = new Set(removedTcertIds);

  await writeJson('attestations', remainingAttestations.filter((a) => !removed.has(a.targetTcertId)));
  await writeJson('pinned', pinned.filter((id) => !removed.has(id)));
  await writeJson('cas', (await readJson<string[]>('cas', [])).filter((id) => !removed.has(id)));
  await writeJson('distrusted', (await readJson<string[]>('distrusted', [])).filter((id) => !removed.has(id)));
  await writeJson('archived-tcerts', (await listArchivedTcertIds()).filter((id) => !removed.has(id)));

  const certificates = await readJson<Array<{ id: string; bytes: string }>>('certificates', []);
  await writeJson('certificates', certificates.filter((cert) => !removed.has(cert.id)));

  const signedByRemoved = <T extends { byKeyId?: string; byTcertId?: string }>(entry: T): boolean =>
    entry.byTcertId === tcertId || (!entry.byTcertId && entry.byKeyId === keyId);
  const revokedTcerts = await readJson<Record<string, RevocationEntry[]>>('revoked-tcerts', {});
  for (const [targetId, entries] of Object.entries(revokedTcerts)) {
    const kept = entries.filter((entry) => !signedByRemoved(entry));
    if (kept.length > 0 && !removed.has(targetId)) revokedTcerts[targetId] = kept;
    else delete revokedTcerts[targetId];
  }
  await writeJson('revoked-tcerts', revokedTcerts);
  const revokedKeys = await readJson<Record<string, RevocationEntry>>('revoked-keys', {});
  for (const [targetId, entry] of Object.entries(revokedKeys)) if (signedByRemoved(entry)) delete revokedKeys[targetId];
  await writeJson('revoked-keys', revokedKeys);
  const revokedAttestations = await readJson<Record<string, RevocationEntry>>('revoked-attestations', {});
  for (const [relation, entry] of Object.entries(revokedAttestations)) {
    const [targetId, caId] = relation.split('|');
    if (caId === tcertId || removed.has(targetId!) || signedByRemoved(entry)) delete revokedAttestations[relation];
  }
  await writeJson('revoked-attestations', revokedAttestations);
  const blocked = await readJson<Record<string, BlockEntry>>('blocked-sdocs', {});
  for (const [sdocId, entry] of Object.entries(blocked)) if (signedByRemoved(entry)) delete blocked[sdocId];
  await writeJson('blocked-sdocs', blocked);
  const sdocStatements = await readJson<Array<{ sdocId: string; entry: SdocStatementEntry }>>('sdoc-statements', []);
  await writeJson('sdoc-statements', sdocStatements.filter(({ entry }) => !signedByRemoved(entry)));

  return { removedTcertIds };
}

/** App-local distribution mirror endpoints per TCert (convenience, not protocol). */
export const endpointConfigStore: IEndpointConfigStore = {
  async getEndpoints(tcertId) {
    return readJson<string[]>(`endpoints.${tcertId}`, []);
  },
  async setEndpoints(tcertId, endpoints) {
    await writeJson(`endpoints.${tcertId}`, [...new Set(endpoints)]);
  },
  async addEndpoint(tcertId, endpoint) {
    const list = await readJson<string[]>(`endpoints.${tcertId}`, []);
    if (!list.includes(endpoint)) {
      list.push(endpoint);
      await writeJson(`endpoints.${tcertId}`, list);
    }
  },
  async removeEndpoint(tcertId, endpoint) {
    const list = await readJson<string[]>(`endpoints.${tcertId}`, []);
    await writeJson(
      `endpoints.${tcertId}`,
      list.filter((e) => e !== endpoint),
    );
  },
};

/** All persisted stores as one object for `createQrsWeb`. */
export const persistedStores = {
  certificateStore,
  documentStore,
  revocationStore,
  trustStore,
  endpointConfigStore,
};

File: qrs-enterprise/enterprise/qrs/stores.py
"""Django-ORM-backed stores implementing the qrs-core-python storage interfaces.

These adapt the qrs-core ``I*Store`` protocols to the enterprise models so the
core's services (certificates, signing, trust, revocation, verification) can run
against the database. Private keys are stored **encrypted at rest** via the
KeyVault (see ``ManagedKey``).
"""
from __future__ import annotations

from typing import Any

from qrs.storage.interfaces import (
    AttestationRecord,
    BlockEntry,
    ICertificateStore,
    IDocumentStore,
    IEndpointConfigStore,
    IPrivateKeyStore,
    IPublicKeyStore,
    IRevocationStore,
    ITrustStore,
    RevocationEntry,
)

from enterprise.models import ManagedKey, ManagedTcert, SdocRecord


class OrmPrivateKeyStore(IPrivateKeyStore):
    """Private keys stored encrypted at rest in ``ManagedKey``."""

    async def save(self, key_id: str, algorithm: str, private_jwk: dict[str, Any]) -> None:
        obj, _ = await _aget_or_create(ManagedKey, key_id=key_id)
        obj.algorithm = algorithm
        obj.set_private_jwk(private_jwk)
        await _asave(obj)

    async def load(self, key_id: str) -> dict[str, Any] | None:
        obj = await _aget(ManagedKey, key_id=key_id)
        if obj is None:
            return None
        private_jwk = obj.get_private_jwk()
        if private_jwk is None:
            return None
        return {"algorithm": obj.algorithm, "private_jwk": private_jwk}

    async def has(self, key_id: str) -> bool:
        return await _aexists(ManagedKey, key_id=key_id)

    async def all(self) -> list[dict[str, Any]]:
        objs = [o async for o in ManagedKey.objects.all()]
        return [
            {"key_id": o.key_id, "algorithm": o.algorithm, "public_jwk": o.public_jwk}
            for o in objs
        ]


class OrmPublicKeyStore(IPublicKeyStore):
    """Public keys stored in ``ManagedKey.public_jwk``."""

    async def save(self, key_id: str, algorithm: str, public_jwk: dict[str, Any]) -> None:
        obj, _ = await _aget_or_create(ManagedKey, key_id=key_id)
        obj.algorithm = algorithm
        obj.public_jwk = public_jwk
        await _asave(obj)

    async def load(self, key_id: str) -> dict[str, Any] | None:
        obj = await _aget(ManagedKey, key_id=key_id)
        if obj is None:
            return None
        return {"algorithm": obj.algorithm, "public_jwk": obj.public_jwk}

    async def has(self, key_id: str) -> bool:
        return await _aexists(ManagedKey, key_id=key_id)

    async def all(self) -> list[dict[str, Any]]:
        objs = [o async for o in ManagedKey.objects.all()]
        return [
            {"key_id": o.key_id, "algorithm": o.algorithm, "public_jwk": o.public_jwk}
            for o in objs
        ]


class OrmCertificateStore(ICertificateStore):
    """TCert bytes stored in ``ManagedTcert.tcert_b64``."""

    async def save(self, tcert_id: str, data: bytes) -> None:
        obj = await _aget(ManagedTcert, tcert_id=tcert_id)
        if obj is not None:
            obj.tcert_b64 = _b64(data)
            await _asave(obj)

    async def get(self, tcert_id: str) -> bytes | None:
        obj = await _aget(ManagedTcert, tcert_id=tcert_id)
        return _unb64(obj.tcert_b64) if obj else None

    async def find_by_key_id(self, key_id: str) -> list[tuple[str, bytes]]:
        objs = [o async for o in ManagedTcert.objects.filter(key__key_id=key_id)]
        return [(o.tcert_id, _unb64(o.tcert_b64)) for o in objs]

    async def all(self) -> list[tuple[str, bytes]]:
        objs = [o async for o in ManagedTcert.objects.all()]
        return [(o.tcert_id, _unb64(o.tcert_b64)) for o in objs]

    async def remove(self, tcert_id: str) -> None:
        await ManagedTcert.objects.filter(tcert_id=tcert_id).adelete()


class OrmDocumentStore(IDocumentStore):
    """SDoc bytes stored in ``SdocRecord.sdoc_b64``."""

    async def save(self, sdoc_id: str, data: bytes) -> None:
        obj, _ = await _aget_or_create(SdocRecord, sdoc_id=sdoc_id)
        obj.sdoc_b64 = _b64(data)
        await _asave(obj)

    async def get(self, sdoc_id: str) -> bytes | None:
        obj = await _aget(SdocRecord, sdoc_id=sdoc_id)
        return _unb64(obj.sdoc_b64) if obj else None

    async def all(self) -> list[tuple[str, bytes]]:
        objs = [o async for o in SdocRecord.objects.all()]
        return [(o.sdoc_id, _unb64(o.sdoc_b64)) for o in objs]

    async def remove(self, sdoc_id: str) -> None:
        await SdocRecord.objects.filter(sdoc_id=sdoc_id).adelete()


class OrmRevocationStore(IRevocationStore):
    """Revocation / block state.

    The enterprise app persists signed statements in ``AuditLog``; the in-memory
    revocation state is rebuilt from the audit trail on demand. For simplicity
    and correctness we keep a lightweight in-memory mirror per runtime, but the
    durable source of truth is the signed statements in the audit log.
    """

    def __init__(self) -> None:
        self._revoked_tcerts: dict[str, list[RevocationEntry]] = {}
        self._revoked_keys: dict[str, list[RevocationEntry]] = {}
        self._blocked_sdocs: dict[str, list[BlockEntry]] = {}

    async def add_revoked_tcert(self, tcert_id: str, entry: RevocationEntry) -> None:
        self._revoked_tcerts.setdefault(tcert_id, []).append(entry)

    async def get_revoked_tcert(self, tcert_id: str) -> RevocationEntry | None:
        entries = self._revoked_tcerts.get(tcert_id) or []
        return entries[-1] if entries else None

    async def list_revoked_tcerts(self) -> list[tuple[str, RevocationEntry]]:
        return [(k, v[-1]) for k, v in self._revoked_tcerts.items() if v]

    async def add_revoked_key(self, key_id: str, entry: RevocationEntry) -> None:
        self._revoked_keys.setdefault(key_id, []).append(entry)

    async def get_revoked_key(self, key_id: str) -> RevocationEntry | None:
        entries = self._revoked_keys.get(key_id) or []
        return entries[-1] if entries else None

    async def list_revoked_keys(self) -> list[tuple[str, RevocationEntry]]:
        return [(k, v[-1]) for k, v in self._revoked_keys.items() if v]

    async def add_blocked_sdoc(self, sdoc_id: str, entry: BlockEntry) -> None:
        self._blocked_sdocs.setdefault(sdoc_id, []).append(entry)

    async def get_blocked_sdoc(self, sdoc_id: str) -> BlockEntry | None:
        entries = self._blocked_sdocs.get(sdoc_id) or []
        return entries[-1] if entries else None

    async def list_blocked_sdocs(self) -> list[tuple[str, BlockEntry]]:
        return [(k, v[-1]) for k, v in self._blocked_sdocs.items() if v]


class OrmTrustStore(ITrustStore):
    """Trust state (pinned / CA / distrusted / attestations).

    Persisted in ``ManagedTcert`` flags plus the audit trail. For the enterprise
    server the trust graph is small; we keep an in-memory mirror rebuilt from the
    database on runtime construction.
    """

    def __init__(self) -> None:
        self._pinned: set[str] = set()
        self._cas: set[str] = set()
        self._distrusted: set[str] = set()
        self._attestations: list[AttestationRecord] = []

    async def add_pinned(self, tcert_id: str) -> None:
        self._pinned.add(tcert_id)

    async def remove_pinned(self, tcert_id: str) -> None:
        self._pinned.discard(tcert_id)

    async def is_pinned(self, tcert_id: str) -> bool:
        return tcert_id in self._pinned

    async def list_pinned(self) -> list[str]:
        return list(self._pinned)

    async def add_ca(self, tcert_id: str) -> None:
        self._cas.add(tcert_id)

    async def remove_ca(self, tcert_id: str) -> None:
        self._cas.discard(tcert_id)

    async def is_ca(self, tcert_id: str) -> bool:
        return tcert_id in self._cas

    async def list_ca(self) -> list[str]:
        return list(self._cas)

    async def add_distrusted(self, tcert_id: str) -> None:
        self._distrusted.add(tcert_id)

    async def remove_distrusted(self, tcert_id: str) -> None:
        self._distrusted.discard(tcert_id)

    async def is_distrusted(self, tcert_id: str) -> bool:
        return tcert_id in self._distrusted

    async def add_attestation(self, record: AttestationRecord) -> None:
        self._attestations.append(record)

    async def get_attestations(self, target_tcert_id: str) -> list[AttestationRecord]:
        return [a for a in self._attestations if a.target_tcert_id == target_tcert_id]

    async def all_attestations(self) -> list[AttestationRecord]:
        return list(self._attestations)


class OrmEndpointConfigStore(IEndpointConfigStore):
    """Endpoint mirrors are stored on ``ManagedTcert.online_endpoint`` (single
    default). Mirrors beyond the signed default are not persisted in v1."""

    async def get_endpoints(self, tcert_id: str) -> list[str]:
        obj = await _aget(ManagedTcert, tcert_id=tcert_id)
        return [obj.online_endpoint] if obj and obj.online_endpoint else []

    async def set_endpoints(self, tcert_id: str, endpoints: list[str]) -> None:
        obj = await _aget(ManagedTcert, tcert_id=tcert_id)
        if obj is not None:
            obj.online_endpoint = endpoints[0] if endpoints else ""
            await _asave(obj)

    async def add_endpoint(self, tcert_id: str, endpoint: str) -> None:
        obj = await _aget(ManagedTcert, tcert_id=tcert_id)
        if obj is not None and not obj.online_endpoint:
            obj.online_endpoint = endpoint
            await _asave(obj)

    async def remove_endpoint(self, tcert_id: str, endpoint: str) -> None:
        obj = await _aget(ManagedTcert, tcert_id=tcert_id)
        if obj is not None and obj.online_endpoint == endpoint:
            obj.online_endpoint = ""
            await _asave(obj)


# ---------------------------------------------------------------------------
# Async ORM helpers (qrs-core services are async)
# ---------------------------------------------------------------------------
def _b64(data: bytes) -> str:
    from qrs.id import to_base64url

    return to_base64url(data)


def _unb64(value: str) -> bytes:
    from qrs.id import from_base64url

    return from_base64url(value)


async def _aget(model, **kwargs):
    try:
        return await model.objects.aget(**kwargs)
    except model.DoesNotExist:
        return None


async def _aget_or_create(model, **kwargs):
    obj = await _aget(model, **kwargs)
    if obj is not None:
        return obj, False
    return model(**kwargs), True


async def _asave(obj) -> None:
    await obj.asave()


async def _aexists(model, **kwargs) -> bool:
    return await model.objects.filter(**kwargs).aexists()